File: langhuan-service/src/main/java/com/langhuan/serviceai/RagService.java
package com.langhuan.serviceai;

import cn.hutool.core.util.IdUtil;
import cn.hutool.json.JSONObject;
import cn.hutool.json.JSONUtil;
import com.langhuan.common.BusinessException;
import com.langhuan.common.Constant;
import com.langhuan.config.VectorStoreConfig;
import com.langhuan.model.domain.TFileUrl;
import com.langhuan.model.domain.TRagFile;
import com.langhuan.service.CacheService;
import com.langhuan.service.MinioService;
import com.langhuan.service.TFileUrlService;
import com.langhuan.service.TRagFileService;
import com.langhuan.dao.VectorStoreRagDao;
import com.langhuan.dao.TFileUrlDao;
import com.langhuan.utils.other.SecurityUtils;
import com.langhuan.utils.rag.EtlPipeline;
import com.langhuan.utils.rag.RagFileVectorUtils;
import com.langhuan.utils.rag.config.SplitConfig;
import jakarta.annotation.Resource;
import lombok.extern.slf4j.Slf4j;
import org.postgresql.util.PGobject;
import org.springframework.ai.document.Document;
import org.springframework.ai.vectorstore.SearchRequest;
import org.springframework.ai.vectorstore.VectorStore;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import static com.langhuan.common.Constant.CACHE_KEY;

@Service
@Slf4j
public class RagService {

    private final TRagFileService ragFileService;
    private final VectorStore ragVectorStore;
    private final ReRankModelService reRankModelService;
    private final EtlPipeline etlPipeline;
    private final TFileUrlService tFileUrlService;
    private final VectorStoreRagDao vectorStoreRagDao;
    private final TFileUrlDao tFileUrlDao;

    @Value("${minio.img-bucket-name}")
    private String bucketName;

    @Resource
    private CacheService cacheService;

    @Resource
    private MinioService minioService;

    public RagService(TRagFileService ragFileService, JdbcTemplate jdbcTemplate,
            VectorStoreConfig vectorStoreConfig, ReRankModelService reRankModelService, EtlPipeline etlPipeline,
            TFileUrlService tFileUrlService, VectorStoreRagDao vectorStoreRagDao, TFileUrlDao tFileUrlDao) {
        this.ragFileService = ragFileService;
        this.ragVectorStore = vectorStoreConfig.ragVectorStore();
        this.reRankModelService = reRankModelService;
        this.etlPipeline = etlPipeline;
        this.tFileUrlService = tFileUrlService;
        this.vectorStoreRagDao = vectorStoreRagDao;
        this.tFileUrlDao = tFileUrlDao;
    }

    @Transactional(rollbackFor = Exception.class)
    public String changeDocumentText(String documents, String documentId, TRagFile ragFile) throws Exception {
        log.info("Updating changeDocumentText: {},documentId: {},documents: {}", ragFile, documentId, documents);
        vectorStoreRagDao.deleteByDocumentId(documentId);
        if (etlPipeline.writeToVectorStore(List.of(documents), etlPipeline.getMetadataFactory().createMetadata(ragFile),
                ragVectorStore)) {
            return "更新成功";
        } else {
            return "更新失败，请检查日志。";
        }
    }

    @Transactional(rollbackFor = Exception.class)
    public String changeDocumentTextByString(String documents, String documentId) throws Exception {
        log.info("Updating changeDocumentTextByString: {}, documentId: {}, documents: {}", documentId, documents);

        List<Map<String, Object>> queryForList = vectorStoreRagDao.selectByDocumentId(documentId);
        PGobject pgObject = (PGobject) queryForList.get(0).get("metadata");
        String string = pgObject.getValue(); // 提取 JSON 字符串
        JSONObject object = JSONUtil.parseObj(string);
        String fileName = object.getStr("filename");
        String fileId = object.getStr("fileId");
        String groupId = object.getStr("groupId");

        vectorStoreRagDao.deleteByDocumentId(documentId);
        if (etlPipeline.writeToVectorStore(List.of(documents),
                etlPipeline.getMetadataFactory().createMetadata(fileName, fileId, groupId),
                ragVectorStore)) {
            return "更新成功";
        } else {
            return "更新失败，请检查日志。";
        }
    }

    @Transactional(rollbackFor = Exception.class)
    public String deleteDocumentText(String documentId, TRagFile ragFile) {
        log.info("Deleting document text: {}, documentId: {}", ragFile, documentId);

        vectorStoreRagDao.deleteByDocumentId(documentId);

        List<TFileUrl> fileUrlList = tFileUrlDao.selectByFileId(ragFile.getId());

        if (fileUrlList != null && !fileUrlList.isEmpty()) {
            deleteImage(fileUrlList);

            int deleteCount = tFileUrlDao.deleteByFileId(ragFile.getId());
            log.info("Deleted {} records from t_file_url", deleteCount);
        }

        Integer newDocNum = Math.max(0, Integer.parseInt(ragFile.getDocumentNum()) - 1);
        ragFile.setDocumentNum(newDocNum.toString());
        boolean updated = ragFileService.updateById(ragFile);
        if (!updated) {
            log.warn("Failed to update ragFile: {}", ragFile.getId());
            throw new RuntimeException("更新文件记录失败");
        }

        return "删除成功";
    }

    public List<String> readAndSplitDocument(MultipartFile file, SplitConfig splitConfig) throws Exception {
        return etlPipeline.process(file, splitConfig);
    }

    public List<String> readAndSplitDocument(String url, SplitConfig splitConfig) throws Exception {
        if (url.startsWith("http")) {
            return etlPipeline.process(url, splitConfig);
        }
        return etlPipeline.process(url, splitConfig, false);

    }

    @Transactional(rollbackFor = Exception.class)
    public String writeDocumentsToVectorStore(List<String> documents, TRagFile ragFile) throws Exception {
        log.info("writeDocumentsToVectorStore: {}", ragFile);

        boolean isInsert = ragFile.getId() == null || ragFile.getId() == 0 || ragFile.getId().toString().isEmpty();
        if (isInsert) {
            ragFile.setId((int) IdUtil.getSnowflakeNextId());
        }
        ragFile.setUploadedAt(new java.util.Date());
        ragFile.setUploadedBy(SecurityUtils.getCurrentUsername());

        // 写入向量库
        boolean writeSuccess = etlPipeline.writeToVectorStore(
                documents,
                etlPipeline.getMetadataFactory().createMetadata(ragFile),
                ragVectorStore);

        if (!writeSuccess) {
            return "添加失败，请检查日志。";
        }

        try {
            // 1. 从缓存中获取 extract 阶段生成的临时 fileId
            Integer tempFileId = cacheService.getId(SecurityUtils.getCurrentUsername() + CACHE_KEY);
            if (tempFileId == null) {
                log.warn("未找到缓存的临时 file_id，跳过 t_file_url 更新");
            } else {
                // 2. 使用 SQL 批量更新 t_file_url 表
                int updatedRows = tFileUrlDao.updateFileIdAndStatus(
                        ragFile.getId(), // 新的 file_id
                        "在用", // 新状态
                        tempFileId, // 旧的临时 file_id
                        "临时" // 原状态
                );

                log.info("批量更新 t_file_url，影响行数: {}, file_id: {} -> {}", updatedRows, tempFileId, ragFile.getId());

                // 3. 清理缓存
                cacheService.removeId(SecurityUtils.getCurrentUsername() + CACHE_KEY);
            }

        } catch (Exception e) {
            log.error("执行 t_file_url 批量更新时出错", e);
            throw new Exception("更新图片引用状态失败", e); // 触发事务回滚
        }

        // 4. 保存或更新 ragFile 记录
        if (isInsert) {
            ragFileService.save(ragFile);
            log.info("保存新文件记录: {}", ragFile.getId());
        } else {
            ragFileService.updateById(ragFile);
            log.info("更新已有文件记录: {}", ragFile.getId());
        }

        return "添加成功";
    }

    @Transactional(rollbackFor = Exception.class)
    public Boolean deleteFileAndDocuments(Integer id) {
        log.info("Deleting file and documents with ID: {}", id);

        int deletedRowsFromVectorStore = vectorStoreRagDao.deleteByFileId(id.toString());
        log.info("Deleted {} rows from vector_store_rag", deletedRowsFromVectorStore);

        List<TFileUrl> fileUrlList = tFileUrlService.lambdaQuery()
                .eq(TFileUrl::getFileId, id)
                .select(TFileUrl::getFUrl, TFileUrl::getId) // 只选择需要的字段
                .list();

        if (fileUrlList != null && !fileUrlList.isEmpty()) {
            deleteImage(fileUrlList);

            List<Integer> idsToDelete = fileUrlList.stream()
                    .map(TFileUrl::getId)
                    .collect(Collectors.toList());
            boolean deletedFromTFileUrl = tFileUrlService.removeByIds(idsToDelete);
            if (!deletedFromTFileUrl) {
                log.warn("Failed to delete records from t_file_url");
                throw new RuntimeException("删除 t_file_url 记录失败");
            }
            log.info("Deleted {} records from t_file_url", idsToDelete.size());
        }

        boolean deletedRagFile = ragFileService.removeById(id);
        if (!deletedRagFile) {
            log.warn("Failed to delete TRagFile record with ID: {}", id);
            throw new RuntimeException("删除 TRagFile 记录失败");
        }

        return true;
    }

    private void deleteImage(List<TFileUrl> fileUrlList) {
        List<String> urlsToDelete = fileUrlList.stream()
                .map(TFileUrl::getFUrl)
                .filter(url -> url != null && !url.trim().isEmpty())
                .toList();

        for (String url : urlsToDelete) {
            try {
                String objectName = minioService.extractObjectName(url, bucketName); // 使用之前的提取方法
                if (objectName != null) {
                    minioService.handleDelete(objectName, bucketName);
                    log.info("MinIO object deleted: {}", objectName);
                }
            } catch (Exception e) {
                log.error("Failed to delete MinIO object: {}", url, e);
                throw new RuntimeException("删除 MinIO 文件失败: " + url, e); // 触发事务回滚
            }
        }
    }

    @Transactional(rollbackFor = Exception.class)
    public String changeFileAndDocuments(TRagFile ragFile) {
        log.info("Updating changeFileAndDocuments: {}", ragFile);
        vectorStoreRagDao.updateGroupIdByFileId(ragFile.getFileGroupId(), ragFile.getId().toString());
        ragFileService.updateById(ragFile);
        return "更新成功";
    }

    public String changeDocumentsRank(String documentId, Integer rank) {
        log.info("Updating changeDocumentsRank: {} rank: {}", documentId, rank);
        vectorStoreRagDao.updateRankByDocumentId(documentId, rank);
        return "更新成功";
    }

    public Map<String, Object> queryDocumentsByFileId(Integer fileId, String content, int pageNum, int pageSize) {
        if (fileId == null) {
            throw new BusinessException("fileId不能为空");
        }
        log.info("queryDocumentsByFileId: {}", fileId);
        return Map.of("list",
                vectorStoreRagDao.selectByFileIdAndContent(fileId.toString(), content, pageSize,
                        (pageNum - 1) * pageSize),
                "count", vectorStoreRagDao.countByFileIdAndContent(fileId.toString(), content));
    }

    public List<Map<String, Object>> queryDocumentsByIds(String fileIds) {
        String[] fileIdsArray = fileIds.split(",");
        if (fileIdsArray.length == 0) {
            throw new BusinessException("fileId不能为空");
        }
        if (fileIdsArray.length > 10) {
            throw new BusinessException("fileId数量不能超过10");
        }

        List<String> documentIds = List.of(fileIdsArray);
        // 执行查询
        log.info("queryDocumentsByFileIds: {}", fileIds);
        List<Map<String, Object>> queryForList = vectorStoreRagDao.selectByIds(documentIds);
        queryForList.forEach(map -> {
            map.put("metadata", map.get("metadata"));
        });
        return queryForList;
    }

    public List<Document> ragSearch(String q, String groupId, String fileId, Boolean isReRank) throws Exception {
        List<Document> searchDocuments = new ArrayList<>();
        isReRank = isReRank == null ? Constant.ISRAGRERANK : isReRank;
        try {
            if (groupId.isEmpty()) {
                searchDocuments = ragVectorStore.similaritySearch(
                        SearchRequest.builder().query(q).topK(Constant.RAGWITHTOPK)
                                .similarityThreshold(Constant.RAGWITHSIMILARITYTHRESHOLD).build() // 单独设置多一些
                );
            } else {
                // HACK:一个组下有相同的文件ID再用 groupId == '" + groupId + "' AND
                String sql = fileId.isEmpty() ? "groupId == '" + groupId + "'" : "fileId == '" + fileId + "'";
                searchDocuments = ragVectorStore.similaritySearch(
                        SearchRequest.builder().query(q).topK(Constant.RAGWITHTOPK)
                                .similarityThreshold(Constant.RAGWITHSIMILARITYTHRESHOLD)
                                .filterExpression(sql)// 设置过滤条件
                                .build());
            }
        } catch (Exception e) {
            log.error("ragSearch error: {}", e.getMessage());
            throw new BusinessException("查询失败");
        }
        if ("linearWeighting".equals(Constant.RAGRANKMODULETYPE)) {
            searchDocuments = rank_linearWeighting(searchDocuments, q, isReRank);
        }
        // 最后结果一定是LLM_RAG_TOPN的数量
        return searchDocuments.subList(0, Math.min(Constant.LLM_RAG_TOPN, searchDocuments.size()));
    }

    /**
     * 线性加权法
     */
    public List<Document> rank_linearWeighting(List<Document> searchDocuments, String q, Boolean isReRank)
            throws Exception {
        if (isReRank) {
            searchDocuments = reRankModelService.chat(q, searchDocuments, searchDocuments.size());
            // 修正归一化排名存储类型
            for (Document doc : searchDocuments) {
                int rank = (int) doc.getMetadata().get("rank");
                doc.getMetadata().put("normalizationRank", rank * 0.01);
            }
        } else {
            // 手工排名归一化指标 加入没relevance_score的情况
            for (Document doc : searchDocuments) {
                int rank = (int) doc.getMetadata().get("rank");
                doc.getMetadata().put("normalizationRank", rank * 0.01);
                doc.getMetadata().put("relevance_score", 1);
            }
        }

        // score（Double类型） -> spring ai的得分 searchDocuments.get(0).get("score");
        // distance（Float类型） -> 数据库向量距离
        // searchDocuments.get(0).getMetadata().get("distance")
        // relevance_score（Double类型） -> rerank模型距离
        // searchDocuments.get(0).getMetadata().get("relevance_score")
        // normalizationRank（Double类型） -> 手工排名
        // searchDocuments.get(0).getMetadata().get("normalizationRank")

        // 线性加权法计算综合得分
        // LINEARWEIGHTING = {数据库向量距离，spring ai得分，rerank模型距离，手工排名}
        for (Document doc : searchDocuments) {
            // 获取各项指标值，安全地处理类型转换
            Object scoreObj = doc.getScore();
            Object distanceObj = doc.getMetadata().get("distance");
            Object rerankScoreObj = doc.getMetadata().get("relevance_score");
            Object normalizedRankObj = doc.getMetadata().get("normalizationRank");

            // 安全转换为Double类型，处理可能的Integer、Float等类型
            Double springAiScore = convertToDouble(scoreObj, 0.0);
            Double vectorDistance = convertToDouble(distanceObj, 1.0);
            Double rerankScore = convertToDouble(rerankScoreObj, 1.0);
            Double normalizedRank = convertToDouble(normalizedRankObj, 0.0);

            // 向量距离需要转换为相似度（距离越小，相似度越高）
            // 这里使用 1 / (1 + distance) 进行转换，确保值在0-1之间
            double distanceSimilarity = 1.0 / (1.0 + vectorDistance);

            // 线性加权计算综合得分
            // 权重顺序：{数据库向量距离，spring ai得分，rerank模型距离，手工排名}
            double weightedScore = Constant.LINEARWEIGHTING[0] * distanceSimilarity +
                    Constant.LINEARWEIGHTING[1] * springAiScore +
                    Constant.LINEARWEIGHTING[2] * rerankScore +
                    Constant.LINEARWEIGHTING[3] * normalizedRank;

            // 将计算出的综合得分存储到metadata中
            doc.getMetadata().put("weightedScore", weightedScore);
        }

        // 根据综合得分降序排序
        searchDocuments.sort((doc1, doc2) -> {
            Double score1 = (Double) doc1.getMetadata().get("weightedScore");
            Double score2 = (Double) doc2.getMetadata().get("weightedScore");
            return Double.compare(score2, score1); // 降序排序
        });

        return searchDocuments;
    }

    /**
     * 安全地将Object转换为Double类型
     * 处理Integer、Float、Double等数值类型的转换
     *
     * @param obj          待转换的对象
     * @param defaultValue 默认值
     * @return 转换后的Double值
     */
    private Double convertToDouble(Object obj, Double defaultValue) {
        if (obj == null) {
            return defaultValue;
        }

        if (obj instanceof Double) {
            return (Double) obj;
        } else if (obj instanceof Integer) {
            return ((Integer) obj).doubleValue();
        } else if (obj instanceof Float) {
            return ((Float) obj).doubleValue();
        } else if (obj instanceof Long) {
            return ((Long) obj).doubleValue();
        } else if (obj instanceof Number) {
            return ((Number) obj).doubleValue();
        } else {
            log.warn("无法转换类型 {} 为Double，使用默认值 {}", obj.getClass().getSimpleName(), defaultValue);
            return defaultValue;
        }

    }
}

File: langhuan-server/src/main/kotlin/com/langhuan/serviceai/RagService.kt
package com.langhuan.serviceai

import cn.hutool.core.util.IdUtil
import cn.hutool.json.JSONUtil
import com.baomidou.mybatisplus.core.conditions.query.QueryWrapper
import com.langhuan.common.BusinessException
import com.langhuan.common.Constant
import com.langhuan.config.VectorStoreConfig
import com.langhuan.dao.TFileUrlDao
import com.langhuan.dao.VectorStoreRagDao
import com.langhuan.model.domain.TFileUrl
import com.langhuan.model.domain.TRagFile
import com.langhuan.model.domain.TRagFileGroup
import com.langhuan.service.*
import com.langhuan.utils.other.SecurityUtils
import com.langhuan.utils.rag.EtlPipeline
import com.langhuan.utils.rag.config.SplitConfig
import jakarta.annotation.Resource
import org.postgresql.util.PGobject
import org.slf4j.LoggerFactory
import org.springframework.ai.document.Document
import org.springframework.ai.vectorstore.VectorStore
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile
import java.util.*

@Service
class RagService(
    vectorStoreConfig: VectorStoreConfig,
    private val ragFileService: TRagFileService,
    private val ragFileGroupService: TRagFileGroupService,
    private val etlPipeline: EtlPipeline,
    private val tFileUrlService: TFileUrlService,
    private val vectorStoreRagDao: VectorStoreRagDao,
    private val tFileUrlDao: TFileUrlDao,
    private val ragCallBackService: RagCallBackService
) {

    companion object {
        private val log = LoggerFactory.getLogger(RagService::class.java)
    }

    private val ragVectorStore: VectorStore = vectorStoreConfig.ragVectorStore()

    @Value("\${minio.img-bucket-name}")
    private var bucketName: String? = null

    @Resource
    private var cacheService: CacheService? = null

    @Resource
    private var minioService: MinioService? = null

    @Transactional(rollbackFor = [Exception::class])
    @Throws(Exception::class)
    fun changeDocumentText(documents: String, documentId: String, ragFile: TRagFile): String {
        log.info("Updating changeDocumentText: {}, documentId: {}, documents: {}", ragFile, documentId, documents)
        vectorStoreRagDao.deleteByDocumentId(documentId)
        return if (etlPipeline.writeToVectorStore(
                listOf(documents), etlPipeline.metadataFactory.createMetadata(ragFile) as Map<String, Any>,
                ragVectorStore
            )
        ) {
            "更新成功"
        } else {
            "更新失败，请检查日志。"
        }
    }

    @Transactional(rollbackFor = [Exception::class])
    @Throws(Exception::class)
    fun changeDocumentTextByString(documents: String, documentId: String): String {
        log.info("Updating changeDocumentTextByString: {}, documentId: {}, documents: {}", documentId, documents)

        val queryForList: List<Map<String, Any>> = vectorStoreRagDao.selectByDocumentId(documentId)
        val pgObject = queryForList[0]["metadata"] as PGobject
        val string = pgObject.value // 提取 JSON 字符串
        val `object` = JSONUtil.parseObj(string)
        val fileName = `object`.getStr("filename")
        val fileId = `object`.getStr("fileId")
        val groupId = `object`.getStr("groupId")

        vectorStoreRagDao.deleteByDocumentId(documentId)
        return if (etlPipeline.writeToVectorStore(
                listOf(documents),
                etlPipeline.metadataFactory.createMetadata(fileName, fileId, groupId),
                ragVectorStore
            )
        ) {
            "更新成功"
        } else {
            "更新失败，请检查日志。"
        }
    }

    @Transactional(rollbackFor = [Exception::class])
    fun deleteDocumentText(documentId: String, ragFile: TRagFile): String {
        log.info("Deleting document text: {}, documentId: {}", ragFile, documentId)

        vectorStoreRagDao.deleteByDocumentId(documentId)

        val fileUrlList = tFileUrlDao.selectByFileId(ragFile.id as Int)

        if (fileUrlList.isNotEmpty()) {
            deleteImage(fileUrlList)

            val deleteCount = tFileUrlDao.deleteByFileId(ragFile.id as Int)
            log.info("Deleted {} records from t_file_url", deleteCount)
        }

        val newDocNum = ragFile.documentNum?.let { maxOf(0, it.toInt() - 1) }
        ragFile.documentNum = newDocNum.toString()
        val updated = ragFileService.updateById(ragFile)
        if (!updated) {
            log.warn("Failed to update ragFile: {}", ragFile.id)
            throw RuntimeException("更新文件记录失败")
        }

        return "删除成功"
    }

    @Throws(Exception::class)
    fun readAndSplitDocument(file: MultipartFile, splitConfig: SplitConfig): List<String> {
        return etlPipeline.process(file, splitConfig)
    }

    @Throws(Exception::class)
    fun readAndSplitDocument(url: String, splitConfig: SplitConfig): List<String> {
        return if (url.startsWith("http")) {
            etlPipeline.process(url, splitConfig)
        } else {
            etlPipeline.process(url, splitConfig, false)
        }
    }

    @Transactional(rollbackFor = [Exception::class])
    @Throws(Exception::class)
    fun writeDocumentsToVectorStore(documents: List<String>, ragFile: TRagFile): String {
        log.info("writeDocumentsToVectorStore: {}", ragFile)

        val isInsert = ragFile.id == null || ragFile.id == 0 || ragFile.id.toString().isEmpty()
        if (isInsert) {
            ragFile.id = IdUtil.getSnowflakeNextId().toInt()
        }
        ragFile.uploadedAt = Date()
        ragFile.uploadedBy = SecurityUtils.getCurrentUsername()

        // 写入向量库
        val writeSuccess = etlPipeline.writeToVectorStore(
            documents,
            etlPipeline.metadataFactory.createMetadata(ragFile) as Map<String, Any>,
            ragVectorStore
        )

        if (!writeSuccess) {
            return "添加失败，请检查日志。"
        }

        try {
            // 1. 从缓存中获取 extract 阶段生成的临时 fileId
            val tempFileId = cacheService?.getId(SecurityUtils.getCurrentUsername() + Constant.CACHE_KEY)
            if (tempFileId == null) {
                log.info("纯文本添加 或 未找到缓存的临时file_id，跳过t_file_url更新")
            } else {
                // 2. 使用 SQL 批量更新 t_file_url 表
                val updatedRows = tFileUrlDao.updateFileIdAndStatus(
                    ragFile.id as Int, // 新的 file_id
                    "在用", // 新状态
                    tempFileId, // 旧的临时 file_id
                    "临时" // 原状态
                )

                log.info("批量更新 t_file_url，影响行数: {}, file_id: {} -> {}", updatedRows, tempFileId, ragFile.id)

                // 3. 清理缓存
                cacheService?.removeId(SecurityUtils.getCurrentUsername() + Constant.CACHE_KEY)
            }

        } catch (e: Exception) {
            log.error("执行 t_file_url 批量更新时出错", e)
            throw Exception("更新图片引用状态失败", e) // 触发事务回滚
        }

        // 4. 保存或更新 ragFile 记录
        if (isInsert) {
            ragFileService.save(ragFile)
            log.info("保存新文件记录: {}", ragFile.id)
        } else {
            ragFileService.updateById(ragFile)
            log.info("更新已有文件记录: {}", ragFile.id)
        }

        return "添加成功"
    }

    @Transactional(rollbackFor = [Exception::class])
    fun deleteFileAndDocuments(id: Int): Boolean {
        log.info("Deleting file and documents with ID: {}", id)

        val deletedRowsFromVectorStore = vectorStoreRagDao.deleteByFileId(id.toString())
        log.info("Deleted {} rows from vector_store_rag", deletedRowsFromVectorStore)

        val fileUrlList = tFileUrlService.query()
            .eq("fileId", id)
            .select("fUrl", "id") // 只选择需要的字段
            .list()

        if (fileUrlList.isNotEmpty()) {
            deleteImage(fileUrlList)

            val idsToDelete = fileUrlList.map { it.id }
            val deletedFromTFileUrl = tFileUrlService.removeByIds(idsToDelete)
            if (!deletedFromTFileUrl) {
                log.warn("Failed to delete records from t_file_url")
                throw RuntimeException("删除 t_file_url 记录失败")
            }
            log.info("Deleted {} records from t_file_url", idsToDelete.size)
        }

        val deletedRagFile = ragFileService.removeById(id)
        if (!deletedRagFile) {
            log.warn("Failed to delete TRagFile record with ID: {}", id)
            throw RuntimeException("删除 TRagFile 记录失败")
        }

        return true
    }

    private fun deleteImage(fileUrlList: List<TFileUrl>) {
        val urlsToDelete = fileUrlList.map { it.fUrl }
            .filter { url: String? -> url != null && url.trim().isNotEmpty() }

        for (url in urlsToDelete) {
            try {
                val objectName = minioService?.extractObjectName(url as String, bucketName as String) // 使用之前的提取方法
                if (objectName != null) {
                    minioService?.handleDelete(objectName, bucketName as String)
                    log.info("MinIO object deleted: {}", objectName)
                }
            } catch (e: Exception) {
                log.error("Failed to delete MinIO object: {}", url, e)
                throw RuntimeException("删除 MinIO 文件失败: $url", e) // 触发事务回滚
            }
        }
    }

    @Transactional(rollbackFor = [Exception::class])
    fun changeFileAndDocuments(ragFile: TRagFile): String {
        log.info("Updating changeFileAndDocuments: {}", ragFile)
        vectorStoreRagDao.updateGroupIdByFileId(ragFile.fileGroupId as String, ragFile.id.toString())
        ragFileService.updateById(ragFile)
        return "更新成功"
    }

    fun changeDocumentsRank(documentId: String, rank: Int): String {
        log.info("Updating changeDocumentsRank: {} rank: {}", documentId, rank)
        vectorStoreRagDao.updateRankByDocumentId(documentId, rank)
        return "更新成功"
    }

    fun queryDocumentsByFileId(fileId: Int?, content: String, pageNum: Int, pageSize: Int): Map<String, Any> {
        if (fileId == null) {
            throw BusinessException("fileId不能为空")
        }
        log.info("queryDocumentsByFileId: {}", fileId)
        return mapOf(
            "list" to vectorStoreRagDao.selectByFileIdAndContent(
                fileId.toString(), content, pageSize,
                (pageNum - 1) * pageSize
            ),
            "count" to vectorStoreRagDao.countByFileIdAndContent(fileId.toString(), content)
        )
    }

    fun queryDocumentsByIds(fileIds: String): List<Map<String, Any>> {
        val fileIdsArray = fileIds.split(",").toTypedArray()
        if (fileIdsArray.isEmpty()) {
            throw BusinessException("fileId不能为空")
        }
        if (fileIdsArray.size > 10) {
            throw BusinessException("fileId数量不能超过10")
        }

        val documentIds = listOf(*fileIdsArray)
        // 执行查询
        log.info("queryDocumentsByFileIds: {}", fileIds)
        val queryForList = vectorStoreRagDao.selectByIds(documentIds)
        return queryForList
    }

    @Throws(Exception::class)
    fun addDocumentToMySpace(documents: List<String>): Boolean {
        val user = SecurityUtils.getCurrentUsername()
        // HACK 和前端约定的知识空间文件组名称
        val setFileGroupName = Constant.DEFAULTFILEGROUPNAME.replace("\${user}", user as String)
        val setFileName = Constant.DEFAULTFILEGROUPFILE.replace("\${user}", user)
        var ragFile = TRagFile()

        val fileGroupList = ragFileGroupService
            .list(QueryWrapper<TRagFileGroup>().eq("group_name", setFileGroupName))

        if (fileGroupList.isEmpty()) {
            val ragFileGroup = TRagFileGroup()
            ragFileGroup.groupName = setFileGroupName
            ragFileGroup.groupType = "个人知识空间"
            ragFileGroup.groupDesc = "对话知识空间生成"
            ragFileGroup.createdBy = user
            ragFileGroup.visibility = "private"
            ragFileGroupService.save(ragFileGroup)
            val fileGroup = ragFileGroupService.getOne(
                QueryWrapper<TRagFileGroup>()
                    .eq("group_name", setFileGroupName)
            )
            ragFile.fileGroupId = fileGroup.id.toString()
        } else {
            // 如果文件组已存在，使用现有文件组的ID
            val existingGroup = fileGroupList[0]
            ragFile.fileGroupId = existingGroup.id.toString()
        }

        // 文件操作
        val fileList = ragFileService.list(QueryWrapper<TRagFile>().eq("file_name", setFileName))
        if (fileList.isEmpty()) {
            ragFile.fileName = setFileName
            ragFile.fileType = "对话知识空间"
            ragFile.fileSize = "无"
            ragFile.documentNum = documents.size.toString()
            ragFile.fileDesc = "$setFileName。在对话中产生。"
            // fileGroupId已经在上面设置过了
            ragFile.uploadedBy = user
            ragFileService.save(ragFile)
        } else {
            val first = fileList.first()
            first.documentNum = first.documentNum?.let { (it.toInt() + documents.size).toString() }
            ragFile = first
            ragFileService.updateById(first)
        }
        val s = writeDocumentsToVectorStore(documents, ragFile)
        return s.contains("成功")
    }

}
